### ISSUE TITLE
Identify and add Type Guards to methods that return a boolean 

### ISSUE THREAD

**Issue by myty** [AndcultureCode/AndcultureCode.JavaScript.Core#73]
For methods that return a `boolean` value, extra type introspection can be given to the typescript compiler and language service by utilizing type guards.  This doesn't apply to all methods that return a `boolean`, but it would be good update those that have been identified as one's that would benefit.

Typescript Docs:
https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

For context, an example, 

```typescript
const hasValue = (value?: string): boolean =>
    // toString is called here to ensure handling all edge cases when a non string value is passed in this function
    value != null && value?.toString().trim() !== "";
```

by adding a type guard to the method,

```typescript
const hasValue = (value?: string): value is string =>
    // toString is called here to ensure handling all edge cases when a non string value is passed in this function
    value != null && value?.toString().trim() !== "";
``` 

any line referencing the string that's run through `hasValue` and is returned as true, the compiler knows that the value is a `string` and not `string | undefined`

**Identified methods:**
- [ ] `StringUtils.hasValue`
- [ ] `StringUtils.isValidEmail`
- [ ] `CollectionUtils.hasValues`
- [ ] `CollectionUtils.isNotEmpty`

**Comment by dtnhinh** [comment]
Can I be assigned to this issue? Thanks

**Comment by myty** [comment]
> Oh, sweet! I think I understand - so for the `CollectionUtils` methods, this will alleviate the need to use the forced non-null operator (`!`) on results that have been passed through it?

Exactly! 👏

**Comment by myty** [comment]
I imagine there could be more instances where we could use this but figured it's a good start with the ones that have been identified so far.

**Comment by dtnhinh** [comment]
Type guard can be done easily for single parameter (i.e. `isEmpty(value?: string)`) but not for rest parameter (`hasValue(...collections: Array<T>`)
![image](https://user-images.githubusercontent.com/69260561/96046401-6bd95380-0e41-11eb-84e2-635278ab9b3f.png)
I am looking to see if there are some user defined type guard for this case

**Comment by brandongregoryscott** [comment]
@dtnhinh Let me know if you find anything that can help narrow down type guards for a rest parameter. If we can't, I'm not opposed to updating the signature of that function to take a single 'collection' - I believe the rest parameter was added to solve some tedious resultObject/API response checking in a complex component for an internal project, but I haven't seen it widely used since. It would be a breaking change, do we feel like the compiler type guard outweighs the rest parameter?

**Comment by dtnhinh** [comment]
I have done some researches but I don't think there is a type guard workaround for rest parameter. I have never seen a case where we parse multiple arguments rather than one single array so I can't say if rest parameter would ever be useful in the future use. Array type guard is very common so let me know if you think we should switch to single collection

**Comment by brandongregoryscott** [comment]
Let's just switch to a single collection @dtnhinh. If we really need one with a rest parameter/multiple collections at once, we can open up a new issue and add in separate function(s) to accomplish that. I think the type guard is a huge advantage for developer ease of use 👍

**Comment by brandongregoryscott** [comment]
Published with [v0.4.0](https://github.com/AndcultureCode/AndcultureCode.JavaScript.Core/releases/tag/v0.4.0)
